### PR TITLE
feat(object-manager): object metadata management based on list-and-watch

### PR DESCRIPTION
**Pull request by Chillax-0v0**

### CODE REVIEW

**Comment by Chillax-0v0** [comment]
TODO: tests

File: components/object-storage/src/object_manager.rs [hunk 1]
```diff
@@ -1,7 +1,408 @@
...
+                    current_offset = end_offset;
+                    objects.push(gen_object_metadata(key, object));
+                    break;
+                }
+            }
+            epoch += 1;
+        }
+
+        // find continuous objects after the first object
+        while epoch <= epoch_max
```
Comment (superhx) [inline components/object-storage/src/object_manager.rs:122]: When there is no more object after the first object, the loop logic will loop 2^16 times.

File: components/object-storage/src/object_manager.rs [hunk 2]
```diff
@@ -1,7 +1,408 @@
...
+        start_offset: key.start_offset,
+        end_offset_delta: object.end_offset_delta,
+        data_len: object.data_len,
+        sparse_index: object.sparse_index.clone(),
+        key: None,
+    }
+}
+
+#[derive(Debug, Default)]
+struct Objects(BTreeMap<ObjectKey, Object>);
```
Comment (superhx) [inline components/object-storage/src/object_manager.rs:66]: Missing part: handle discontinuous committed object

**Review by lizhanhui — APPROVED** [review]

File: components/object-storage/src/object_manager.rs (hunk 1 above)
Comment (Chillax-0v0) [reply components/object-storage/src/object_manager.rs:122]: `epoch_max` is default to `u16::MIN`, so the loop will break fast.

File: components/object-storage/src/range_offload.rs [hunk 3]
```diff
@@ -407,7 +408,7 @@ where
             bytes.extend_from_slice(&gen_footer(payload_length as u32, sparse_index_len as u32));
             let bytes = bytes.freeze();
             Self::write_object(&op, &key, &bytes).await;
-            object_manager.commit_object(object_metadata);
+            _ = object_manager.commit_object(object_metadata).await;
```
[the remark this answers is not in the document]
Comment (Chillax-0v0) [reply components/object-storage/src/range_offload.rs:410]: Fix

**Pull request merged**